clust/utils.q

```diff
@@ -0,0 +1,1323 @@
+// clust/utils.q - Clustering Utilities
+// Copyright (c) 2021 Kx Systems Inc
+// 
+// Collection of utility functions for 
+// implimentation of clustering algos
+
+\d .ml
+
+
+// Distance metric dictionary
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Euclidean distance calculation
+// @param data {float[][]} Points
+// @return {float[]} Euclidean distances for data 
+clust.i.df.edist:{[data]
+  sqrt data wsum data
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Distance calculation
+// @param data {float[][]} Points
+// @return {float[]} Euclidean squared distances for data 
+clust.i.df.e2dist:{[data]
+  data wsum data
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Manhattan distance calculation
+// @param data {float[][]} Points
+// @return {float[]} Manhattan distances for data 
+clust.i.df.mdist:{[data]
+  sum abs data
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Chebyshev distance calculation
+// @param data {float[][]} Points
+// @return {float[]} Chebyshev distances for data 
+clust.i.df.cshev:{[data]
+  min abs data
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Negative euclidean squared distance calculation
+// @param data {float[][]} Points
+// @return {float[]} Negative euclidean squared distances for data 
+clust.i.df.nege2dist:{[data]
+  neg data wsum data
+  }
+
+// @private
+// @kind dictionary
+// @category clustUtility
+// @desc Linkage dictionary
+// @type dictionary
+clust.i.lf.single:min
+clust.i.lf.complete:max
+clust.i.lf.average:avg
+clust.i.lf.centroid:raze
+clust.i.lf.ward:{z*x*y%x+y}
+
+// Distance calculations
+
+// @private
+// @kind function
+// @category clustUtility
+// @param data {float[][]} Points in `value flip` format
+// @param df {fn} Distance function
+// @param pt {float[]} Current point
+// @param idxs {long[]} Indices from data
+// @return {float[]} Distances for data and pt
+clust.i.dists:{[data;df;pt;idxs]
+  clust.i.df[df]pt-data[;idxs]
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @param data {float[][]} Points in `value flip` format
+// @param df {fn} Distance function
+// @param pt {float[]} Current point
+// @param idxs {long[]} Indices from data
+// @return {float[]} Distances for data and pt
```

Comment:
This doesn't return a float list, it returns a dictionary indicating the index of the closest point and the distance between that point and the point on interest `current point`